crates/interpreter/src/exec.rs

```diff
@@ -773,28 +789,44 @@ impl<'m> Thread<'m> {
             Unreachable => return Err(trap()),
             Nop => (),
             Block(b) => self.push_label(self.blocktype(inst, &b), LabelKind::Block),
-            Loop(b) => self.push_label(self.blocktype(inst, &b), LabelKind::Loop(saved)),
+            Loop(b) => {
+                let side_table_idx = self.frame().side_table_indices.1;
+                self.push_label(
+                    self.blocktype(inst, &b),
+                    LabelKind::Loop(LoopState { parser_data: saved, side_table_idx }),
+                )
+            }
             If(b) => match self.pop_value().unwrap_i32() {
                 0 => {
-                    self.skip_to_else(inst);
+                    self.take_jump(None, inst.module.side_tables());
                     self.push_label(self.blocktype(inst, &b), LabelKind::Block);
                 }
-                _ => self.push_label(self.blocktype(inst, &b), LabelKind::If),
+                _ => {
+                    self.frame().skip_jump();
+                    self.push_label(self.blocktype(inst, &b), LabelKind::If);
+                }
             },
             Else => {
-                self.skip_to_end(inst, 0);
+                self.take_jump(None, inst.module.side_tables());
                 return Ok(self.exit_label());
             }
-            End => return Ok(self.exit_label()),
-            Br(l) => return Ok(self.pop_label(inst, l)),
+            End => {
+                return Ok(self.exit_label());
+            }
+            Br(l) => return Ok(self.pop_label(inst, l, None)),
             BrIf(l) => {
                 if self.pop_value().unwrap_i32() != 0 {
-                    return Ok(self.pop_label(inst, l));
+                    return Ok(self.pop_label(inst, l, None));
                 }
+                self.frame().skip_jump();
             }
             BrTable(ls, ln) => {
                 let i = self.pop_value().unwrap_i32() as usize;
-                return Ok(self.pop_label(inst, ls.get(i).cloned().unwrap_or(ln)));
+                return Ok(self.pop_label(
+                    inst,
+                    ls.get(i).cloned().unwrap_or(ln),
+                    ls.get(i).map(|_| i),
```

Comment:
Actually, if we make this `ls.get(i).map_or(0, |_| i + 1)` then we don't need an `Option` and we can use `0` instead of `None` for the other calls. This should also improve performances a bit by corner-casing only the BrTable.

Reply:
Very clever idea!